1 1 java 14 features

hontardmitry opened the pull request.

hontardmitry committed:
hometask 1 1
initial commit

hontardmitry committed:
hometask 1 1
tests added
Logger added
annotations added

hontardmitry committed:
hometask 1 1
parametrizing the test

hontardmitry committed:
reorganize folder

hontardmitry committed:
comments deleted

hontardmitry committed:
cleaned up

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Test
+    @DisplayName("Regular user login test")
+    public void checkSuccessfulLogin(TestInfo testInfo) {
```
Remove unused "testInfo" value from parameter of this test, please

hontardmitry replied:
Done
Thank you! I've forgotten that moved using of TestInfo to BaseTest

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+    private final static String BASE_URL = getBaseUrl();
+    private final static String INVENTORY_URL = getInventoryUrl();
+    private static final String STANDARD_USER = getStandardUser();
+    private static final String STANDARD_USER_PWD = getStandardUserPwd();
```
Create an enum and put all these values ​​into it, please. We will be able to use these values ​​directly from it. So, we won't need these steps (from 31 to 34 lines) in this class

hontardmitry replied:
Thank you, I've created an enum with these values. It looks prettier now, but I doubt if the current implementation is not ugly. Please, share your thoughts.

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+import java.util.Map;
+
+public class SauceTest extends BaseTest {
```
All our tests should be located in "test" directory. But now we're having just "java" directory, and all our classes(and even  
test runner) are located here. So, could you create following hierarchy in "src" package: test -> java(mark it like Test Sources Root) -> tests; And put this class into it.

P.S. I guess you meant to name it "SourceTest", not "SauceTest")))

hontardmitry replied:
Ok, I've done these hierarchy changes. And I have several questions in regards to this point which I'd like to discuss verbally.

P.S. SauceTest inherits its name from the site name I run tests over: "www.saucedemo.com"

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.Map;
+
+public class SauceTest extends BaseTest {
+    private static final Logger LOGGER = getLoggerForCurrentClass();
```
You don't use log4j logger here. Even in "imports" you can find "import org.slf4j.Logger".
So, if you want to use log4j, you need "import org.apache.log4j.Logger;" import and initializing of logger instance would look like "private static final Logger LOGGER = Logger.getLogger(SauceTest.class);";

hontardmitry replied:
log4j dependency is removed from pom.xml

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+public class SauceTest extends BaseTest {
+    private static final Logger LOGGER = getLoggerForCurrentClass();
+
+    private final static PageHeader pageHeader = new PageHeader();
```
Move it into the "BaseTest" class, please

hontardmitry replied:
Done

AntBelyi commented on hometask1_1-java_14-features/src/main/java/utils/LogUtils.java:
```diff
@@ -0,0 +1,13 @@
+        final String clazz = t.getStackTrace()[1].getClassName();
+        return LoggerFactory.getLogger(clazz);
+    }
+}
```
If you will switch to log4j you will not need this class with this functionality.

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+    @DisplayName("Navigation menu options test")
+    public void checkNavigationMenuOptions(TestInfo testInfo) {
+        loginWithCredentials(STANDARD_USER, STANDARD_USER_PWD);
+        pageHeader.getNavigationMenuButton().should(Condition.visible).click();
```
Move this "should(Condition.visible)" weiter into the "getNavigationMenuButton()" method, please.

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Test
+    @DisplayName("Navigation menu options test")
+    public void checkNavigationMenuOptions(TestInfo testInfo) {
```
1) This test is failed. Fix it please.
2) Remove unused "testInfo" value from parameter of this test, please

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Test
+    @DisplayName("Null pointer exception test")
+    public void checkNullPointer(TestInfo testInfo) {
```
Remove unused "testInfo" value from parameter of this test, please

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+    @DisplayName("Null pointer exception test")
+    public void checkNullPointer(TestInfo testInfo) {
+        loginWithCredentials(STANDARD_USER, STANDARD_USER_PWD);
+        pageHeader.getCartButton().should(Condition.visible).click();
```
Move this "should(Condition.visible)" weiter into the "getNavigationMenuButton()" method, please.

hontardmitry replied:
Good point. Moved "should(Condition.visible)" to get selenide element methods

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+        ElementsCollection filteredList = null;
+
+        LOGGER.info("Provoke NPE");
+        assertTrue(filteredList.get(0).$(".inventory_item_price").exists());
```
What do I need to do to get the description of ""NullPointerException" that java 14 provides us with?

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/tests/SauceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+        assertEquals(expectedStatus, actualStatus);
+    }
+}
```
It's something like unit test but not ui. Write UI test, and try to apply this functionality to it.

AntBelyi commented on hometask1_1-java_14-features/src/main/java/utils/Waiter.java:
```diff
@@ -0,0 +1,16 @@
+        WebDriverWait waiter = new WebDriverWait(getWebDriver(), 10000);
+        waiter.until(driver -> "complete".equals(
+            executeJavaScript("return document.readyState").toString()));
+    }
```
You use here two waiters from different frameworks simultaneously. Why? For example on the line 11 you use selenide "sleep(...)", on the line 12, 13, 14 it's selenium waiter. I guess If we use selenide - we need to use it's functionality.
And to be honest, I'm not sure that we need this method. You will be able just to use something like "should(Condition.visible)" in "loginWithCredentials(...)" method. Just choose any element on the page(after login page), and wait it.

hontardmitry replied:
Indeed, this method is redundant here. And it seems that I should refactor this method in our test framework on the project as well to make it use a single framework :). Actually, I didn't pay enough attention to the way how it was implemented.

AntBelyi commented on .idea/.name:
```diff
@@ -0,0 +1 @@
+hometask1_1
```
Try to ignore all these useless files

hontardmitry committed:
first part of fixes

hontardmitry committed:
Switch-case ui test

hontardmitry committed:
Switch-case ui test

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/pages/components/PageFooter.java:
```diff
@@ -0,0 +1,14 @@
+
+public class PageFooter {
+    public ElementsCollection getSocialLinks(){
+        $(".social").should(Condition.visible);
```
Avoid this hardcoding please. It isn't clear, what this element is.

Create variable and then reuse it here. For example:
private SelenideElement 'nameOfElement' = .......

And for the lower element the same

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/pages/components/PageHeader.java:
```diff
@@ -0,0 +1,29 @@
+
+        return $(".shopping_cart_link").should(Condition.visible);
+    }
+}
```
Try to avoid hardcoding here please. Create private variables for all elements, and reuse them in all methods

AntBelyi commented on hometask1_1-java_14-features/src/main/java/ui/pages/LoginPage.java:
```diff
@@ -0,0 +1,23 @@
+    public SelenideElement getLoginButton() {
+
+        return $("#login-button").should(Condition.visible);
+    }
```
02.09 String variables using approach was changed to SelenideElements using

AntBelyi commented on hometask1_1-java_14-features/src/test/java/ui/SauceTest.java:
```diff
@@ -0,0 +1,109 @@
+            default -> throw new InputMismatchException("Unexpected class provided");
+        };
+        assertEquals(expectedLink, link.$("a").getAttribute("href"));
+    }
```
we don't need to store such methods in "TestRunner" class.
Think, where could you move it

hontardmitry committed:
Got rid of hardcoded locators
Moved assertion to a separate class
Got rid of using of logger factory from slf4j
slf4j-log4j12 dependency is needed for selenide to avoid SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder" error

hontardmitry committed:
instanceof dummy example added

hontardmitry committed:
Merge branch 'main' into 1_1-java_14_features

hontardmitry committed:
change Strings to SelenideElements in Pages classes

hontardmitry commented:
> 

String variables were replaced with SelenideElement variables
